pybossa/cache/projects.py

```diff
@@ -174,42 +160,91 @@ def format_task(row, lock_users=[]):
+        sql_unlock_filter = " AND id NOT IN ({})".format(",".join(locked_task_ids))
+        sql_order = " ORDER BY {} "
+        sql_limit_offset = " LIMIT :limit OFFSET :offset "
+        params["limit"] = limit
+        params["offset"] = offset
+
+        if "lock_status" in order_by:
+            sql_order_by = sorting[order_by]
+
+            # if there are locked tasks in project, need to merge sort locked tasks and unlocked tasks
+            # otherwise just sort by pcomplete
+            if locked_tasks_in_project:
+                locked_tasks = [format_task(row, locked_tasks_in_project.get(row.id, []))
+                                    for row in session.execute(text(sql+sql_lock_filter), params)]
+
+                if order_by == 'lock_status asc':
+                    # sort by completed tasks, then incomplete tasks, then locked tasks
+
+                    # if current offset is less than number of unlocked tasks
+                    #   need to query unlocked tasks first, then append locked tasks to fulfill the page size
+                    # if current offset is more than number of unlocked tasks
+                    #   should skip all unlocked tasks and partial locked tasks, and present the rest locked tasks
+                    if offset < total_count - len(locked_tasks):
+                        sql_query = sql + sql_unlock_filter + sql_order.format(sql_order_by) + sql_limit_offset
+                        results = session.execute(text(sql_query), params)
+                    else:
+                        locked_tasks = locked_tasks[offset-(total_count-len(locked_tasks)):]
+                        results = []
+
+                    for row in results:
+                        tasks.append(format_task(row, locked_tasks_in_project.get(row.id, [])))
+
+                    # fill up the page size with locked tasks until the page size is reached or exhausted
+                    for lt in locked_tasks:
+                        if len(tasks) < limit:
+                            tasks.append(lt)
+                        else:
+                            break
+
+                else:
+                    # sort by locked tasks, then incompleted tasks, then completed tasks
+                    tasks = locked_tasks[offset: offset+limit]
+
+                    # if locked tasks is not enough to fulfill the page,
+                    #   need to append unlocked tasks until the page size is fulfilled
+                    #   query additional unlocked tasks and append them until the page size is reached or exhausted
+                    if len(tasks) < limit:
+                        sql_query = sql + sql_unlock_filter + sql_order.format(sql_order_by) + sql_limit_offset
+                        params["offset"] = max(params["offset"]-len(locked_tasks), 0)
+                        results = session.execute(text(sql_query), params)
+
+                        for row in results:
```

Comment:
Instead of appending one by one until it fulfills the page, can we construct a SQL with "limit" clause, where the "limit" count is "limit - len(tasks)"?

Reply:
Good point, will update.